doc/topics/blocking-vs-non-blocking.md

```diff
@@ -0,0 +1,140 @@
+# Overview of Blocking vs Non-Blocking
+
+This overview covers the difference between blocking and non-blocking calls in
+Node.js. This overview will refer to the event loop and libuv but no prior
+knowledge of those topics is required. Readers are assumed to have a basic
+understanding of the JavaScript language and Node.js callback pattern.
+
+> In this overview, "I/O" refers primarily to interaction with the system's disk
+and network supported by [libuv](http://libuv.org/).
+
+
+## Blocking
+
+Blocking is when the execution of additional JavaScript in the Node.js process
+must wait until an I/O operation completes. Blocking may occur when using any of
+the synchronous I/O methods in the Node.js standard library that use libuv. Use
+of blocking methods prevents the event loop from doing additional work while
+waiting for I/O to complete.
+
+In Node.js, JavaScript that exhibits poor performance due to being CPU intensive
+rather than I/O bound, isn't typically referred to as blocking.
+
+All of the I/O methods in the Node.js standard libraries provide asynchronous
+versions, which are non-blocking, and accept callback functions. Some methods
+also implement blocking methods, which usually have names that end with `Sync`.
+
+
+## Comparing Code
+
+Using the File System module as an example, this is a synchronous file read:
+
+```js
+const fs = require('fs');
+const data = fs.readFileSync('/file.md'); // blocks here until file is read
+```
+
+And here is an equivalent asynchronous example:
+
+```js
+const fs = require('fs');
+fs.readFile('/file.md', (err, data) => {
+  if (err) throw err;
+});
+```
+
+The first example appears simpler but it has the disadvantage of the second line
+blocking the execution of any additional JavaScript until the entire file is
+read. Note that in the synchronous version if an error is thrown it will need to
+be caught or the process will crash. In the asynchronous version, it is up to
+the author to decide whether an error should throw as shown.
+
+Let's expand our example a little bit:
+
+```js
+const fs = require('fs');
+const data = fs.readFileSync('/file.md'); // blocks here until file is read
+console.log(data);
+moreWork(); // will run after console.log
+```
+
+And here is a similar, but not equivalent asynchronous example:
+
+```js
+const fs = require('fs');
+fs.readFile('/file.md', (err, data) => {
+  if (err) throw err;
+  console.log(data);
+});
+moreWork(); // will run before console.log
+```
+
+In the first example above, `console.log` will be called before `moreWork`. In
+the second example `readFile` is non-blocking so JavaScript execution can
+continue and `moreWork` will be called first. The ability to run `moreWork`
```

Comment:
question - are we always guaranteed that `readFile` will execute its callback in a following turn of the event loop? I assume yes since it has to hit libuv at least once but I want to be sure.